### ISSUE TITLE
Is there a way to know the pre-commit run on the whole repo (flag `--all`/`-a` provided)?

### ISSUE THREAD

**Issue by MaxymVlasov** [pre-commit/pre-commit#2556]
Hi,

Some hooks execute much more time when executes on each dir, than when executed once on the whole repo.
We already have a workaround for small repos based on [git ls-files](https://github.com/antonbabenko/pre-commit-terraform/blob/07c6764c309d14fd37536e6c679354ab99c59a59/hooks/_common.sh?rgh-link-date=2022-10-22T08%3A19%3A50Z#L160-L164), but that doesn't work for big repos, because files get partitioned by [xargs from pre-commit](https://github.com/pre-commit/pre-commit/blob/0239b27f4ffb37d56b94ac3f9aecef22300daf19/pre_commit/languages/helpers.py#L132).

So, is there a way to know the pre-commit run on the whole repo (flag `--all`/`-a` provided)? Eg, by env variable, etc.

Based on https://github.com/antonbabenko/pre-commit-terraform/issues/443, originally submitted by @anrub

**Comment by asottile** [comment]
there is not (and won't be implemented), and there shouldn't be a performance difference if done correctly

**Comment by asottile** [comment]
there's actually a few dupes -- please follow the template and search the tracker next time

**Comment by MaxymVlasov** [comment]
Double check issues, and if specified too specific - there are no related issues, if specify too open search - it's about 31 pages...
Can you please point me to duplicates? Maybe I have no idea how to rephrase it to find exactly what I want.

**Comment by MaxymVlasov** [comment]
By the way
![image](https://user-images.githubusercontent.com/11096782/197587888-57fc2957-2b8e-426a-a7cc-989294bf2b49.png)

**Comment by yermulnik** [comment]
> and there shouldn't be a performance difference if done correctly

@asottile Do you mean `pre-commit` has been designed the way that running self-contained binary as a shell-based hook via `pre-commit` framework locally to executor multiple times (even in parallel) should have no notable performance impact compared to running it only once within the same conditions? Thanks.

**Comment by asottile** [comment]
yes, when a tool follows the framework -- linting a sequence of positional argument filenames

**Comment by yermulnik** [comment]
> linting a sequence of positional argument filenames

Given a repo with e.g. 10K files updated within single commit, is your statement still valid?

**Comment by asottile** [comment]
what are you trying to get at?  just spit it out instead of speaking in riddles

**Comment by yermulnik** [comment]
I'm just trying to figure whether there's a reason for `pre-commit` to withhold giving info to underlying tools about `--all-files` parameter being passed by caller as we have a use case where a user reports notable time performance issue when running `pre-commit` over a pretty much large repo as compared to running an underlying tool once.

**Comment by asottile** [comment]
yes, generally `xargs -P n tool ...` is faster than `xargs tool ...`

giving tools information about `--all-files` introduces a mechanism to have *surprising behaviour* and if tools are written correctly they won't benefit from it

**Comment by yermulnik** [comment]
> yes, generally `xargs -P n tool ...` is faster than `xargs tool ...`

yes, definitely. but I'm talking about no `xargs` at all:
```bash
> sync

> find ./ | wc -l
10051

> git ls-files | wc -l
10004

> time terragrunt hclfmt ./

real    0m0,235s
user    0m0,256s
sys     0m0,064s

> time find ./ -print0 | xargs -P $(egrep -c ^processor /proc/cpuinfo) -0 terragrunt hclfmt

real    0m0,396s
user    0m0,684s
sys     0m0,172s
```

And user reports this when triggering `terragrunt hclfmt` via our shell-based hook:
![image](https://user-images.githubusercontent.com/1274789/197895795-9b4cf93d-6c7e-4e7c-b8ca-6dffb64d5de2.png)
So we're trying to figure whether it might be possible to run hook once instead of x`xargs` times 🤔 
Would appreciate your input on where we could have introduced hook misbehavoiur. Thanks.

**Comment by asottile** [comment]
for example, linting files which aren't part of the repository

your shell is slow because it doesn't follow the normal way of doing this (it "dedupes" to dirs and then will do significant duplicated work)

and [never post images of text](https://youtu.be/svlsnZFWWDE)
